docs/docs/dev_docs/contracts/resources/common_patterns/authwit.md

```diff
@@ -0,0 +1,224 @@
+For private calls where we allow execution on behalf of others, we generally want to check if the current call is authenticated by `on_behalf_of`. To easily do so, we can use the `assert_current_call_valid_authwit` which fetches information from the current context without us needing to provide much beyond the `on_behalf_of`.
+
+#include_code assert_current_call_valid_authwit /yarn-project/aztec-nr/authwit/src/auth.nr rust
+
+As seen above, we mainly compute the message hash, and then forward the call to the more generic `assert_valid_authwit`. This validating function will then:
+
+- make a call to `on_behalf_of` to validate that the call is authenticated
+- emit a nullifier for the action to prevent replay attacks
+- throw if the action is not authenticated by `on_behalf_of`
+
+#include_code assert_valid_authwit /yarn-project/aztec-nr/authwit/src/auth.nr rust
+
+### Utilities for public calls
+
+Very similar to above, we have variations that work in the public domain. These functions are wrapped to give a similar flow for both cases, but behind the scenes the logic of the account contracts is slightly different since they cannot use the oracle as they are not in the private domain.
+
+#include_code assert_current_call_valid_authwit_public /yarn-project/aztec-nr/authwit/src/auth.nr rust
+
+#include_code assert_valid_authwit_public /yarn-project/aztec-nr/authwit/src/auth.nr rust
+
+## Usage
+
+Ok, enough talking, how the hell do we use this?
+
+### Importing it
+
+To add it to your project, add the `authwit` library to your `Nargo.toml` file.
+
+```toml
+[dependencies]
+aztec = { git="https://github.com/AztecProtocol/aztec-packages/", tag="#include_aztec_version", directory="yarn-project/aztec-nr/aztec" }
+authwit = { git="https://github.com/AztecProtocol/aztec-packages/", tag="#include_aztec_version", directory="yarn-project/aztec-nr/authwit"}
+```
+
+Then you will be able to import it into your contracts as follows.
+
+#include_code import_authwit /yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr rust
+
+### Private Functions
+
+#### Checking if the current call is authenticated
+
+Based on the diagram earlier on this page let's take a look at how we can implement the `transfer` function such that it checks if the tokens are to be transferred `from` the caller or needs to be authenticated with an authentication witness.
+
+#include_code transfer /yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr rust
+
+The first thing we see in the snippet above, is that if `from` is not the call we are calling the `assert_current_call_valid_authwit` function from [earlier](#private-functions). If the call is not throwing, we are all good and can continue with the transfer.
+
+In the snippet we are constraining the `else` case such that only `nonce = 0` is supported. This is not strictly necessary, but because I can't stand dangling useless values. By making it constrained, we can limit what people guess it does, I hope.
```

Comment:
:laughing:, but it feels odd for the "I" to suddenly pop up in the middle of the doc